File: 2022/day-4/day-4.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	flag.Parse()
	bytes, err := ioutil.ReadFile("input.txt")
	if err != nil {
		return
	}
	contents := string(bytes)
	split := strings.Split(contents, "\n")

	var part1, part2 int
	for _, s := range split[:len(split)] {
		pairs := strings.Split(s, ",")
		left := strings.Split(pairs[0], "-")
		right := strings.Split(pairs[1], "-")
		leftStart, _ := strconv.Atoi(left[0])
		leftEnd, _ := strconv.Atoi(left[1])
		rightStart, _ := strconv.Atoi(right[0])
		rightEnd, _ := strconv.Atoi(right[1])
		if leftStart >= rightStart && leftEnd <= rightEnd {
			part1++
			part2++
			continue
		}
		if leftStart <= rightStart && leftEnd >= rightEnd {
			part1++
			part2++
			continue
		}
		if leftStart >= rightStart && leftStart <= rightEnd {
			part2++
			continue
		}
		if rightStart >= leftStart && rightStart <= leftEnd {
			part2++
			continue
		}
		if leftEnd >= rightStart && leftEnd <= rightEnd {
			part2++
			continue
		}
		if rightEnd >= leftStart && rightEnd <= leftEnd {
			part2++
			continue
		}
	}

	fmt.Println("Part 1 solution:", part1)
	fmt.Println("Part 2 solution:", part2)
}

File: 2022/day-4/day-4.py
with open("input.txt", "r") as f:
    lines = f.readlines()
    pairs = [line.strip() for line in lines]

def createSets(pairs):
    left = pairs.split(",")[0]
    left_start = int(left.split("-")[0].strip())
    left_stop = int(left.split("-")[1].strip())
    left_range = range(left_start, left_stop+1)

    right = pairs.split(",")[1]
    left_stop = int(right.split("-")[0].strip())
    right_stop = int(right.split("-")[1].strip())
    right_range = range(left_stop, right_stop+1)

    return left_range, right_range

def part1(pairs):
    counter = 0

    for i in pairs:
        left_range, right_range = createSets(i)
        left_set = set(left_range)
        right_set = set(right_range)

        if left_set.issubset(right_set) or right_set.issubset(left_set):
            counter +=1

    return counter

def part2(pairs):
    counter = 0

    for i in pairs:
        left_range, right_range = createSets(i)
        left_set = set(left_range)
        right_set = set(right_range)
        overlap = left_set & right_set

        if len(overlap)>0:
            counter+=1

    return counter

if __name__ == "__main__":
    print('Part 1 solution:', part1(pairs))
    print('Part 2 solution:', part2(pairs))
